File: astring/src/test/java/xyz/tynn/astring/AStringKtTest.java
//  Copyright 2020 Christian Schmitz
//  SPDX-License-Identifier: Apache-2.0

package xyz.tynn.astring;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import static io.mockk.MockKKt.every;

import android.content.Context;
import android.view.View;

import androidx.fragment.app.Fragment;

import org.junit.Rule;
import org.junit.Test;

import io.mockk.impl.annotations.MockK;
import io.mockk.impl.annotations.RelaxedMockK;
import io.mockk.junit4.MockKRule;

public class AStringKtTest {

    @Rule
    public final MockKRule mockkRule = new MockKRule(this);

    @RelaxedMockK
    AString aString;

    @MockK
    Context context;
    @RelaxedMockK
    Fragment fragment;
    @RelaxedMockK
    View view;

    @Test
    public void invokeWithContext_should_delegate_to_context() {
        every(scope -> aString.invoke(context)).returns("context");
        assertEquals("context", AStringKt.invokeWithContext(context, aString));
    }

    @Test
    public void invokeWithContext_should_delegate_null_to_context() {
        assertNull(AStringKt.invokeWithContext(context, null));
    }

    @SuppressWarnings("ConstantConditions")
    @Test(expected = NullPointerException.class)
    public void invokeWithContext_should_throw_on_null_context() {
        AStringKt.invokeWithContext((Context) null, aString);
    }

    @Test
    public void invokeWithFragment_should_delegate_to_fragment() {
        every(scope -> aString.invoke(fragment.requireContext())).returns("fragment");

        assertEquals("fragment", AStringKt.invokeWithFragment(fragment, aString));
    }

    @Test
    public void invokeWithFragment_should_delegate_null_to_fragment() {
        assertNull(AStringKt.invokeWithFragment(fragment, null));
    }

    @SuppressWarnings("ConstantConditions")
    @Test(expected = NullPointerException.class)
    public void invokeWithFragment_should_throw_on_null_fragment() {
        AStringKt.invokeWithFragment(null, aString);
    }

    @Test(expected = IllegalStateException.class)
    public void invokeWithFragment_should_throw_on_null_context() {
        every(scope -> fragment.requireContext()).answers((scope, $) -> scope.callOriginal());
        every(scope -> fragment.getContext()).returns(null);
        AStringKt.invokeWithFragment(fragment, aString);
    }

    @Test
    public void invokeWithView_should_delegate_to_view() {
        every(scope -> aString.invoke(view.getContext())).returns("view");
        assertEquals("view", AStringKt.invokeWithView(view, aString));
    }

    @Test
    public void invokeWithView_should_delegate_null_to_view() {
        assertNull(AStringKt.invokeWithView(view, null));
    }

    @SuppressWarnings("ConstantConditions")
    @Test(expected = NullPointerException.class)
    public void invokeWithView_should_throw_on_null_view() {
        AStringKt.invokeWithView((View) null, aString);
    }
}

File: astring/src/test/kotlin/xyz/tynn/astring/AStringKtTest.kt
//  Copyright 2020 Christian Schmitz
//  SPDX-License-Identifier: Apache-2.0

package xyz.tynn.astring

import android.app.Activity
import android.content.Context
import android.view.View
import androidx.fragment.app.Fragment
import io.mockk.every
import io.mockk.mockk
import io.mockk.spyk
import io.mockk.verify
import kotlin.test.Test

internal class AStringKtTest {

    val aString = mockk<AString>(relaxed = true)
    val aContext = mockk<Context>()

    @Test
    fun `invoke with fragment delegates to context`() {
        aString(mockk<Fragment> {
            every { requireContext() } returns aContext
        })

        verify { aString(aContext) }
    }

    @Test
    fun `invoke with view delegates to context`() {
        aString(mockk<View> {
            every { context } returns aContext
        })

        verify { aString(aContext) }
    }

    @Test
    fun `getValue with context delegates to this`() {
        val aContext = object : Activity() {
            val string by aString
        }.apply { string }

        verify { aString(aContext) }
    }

    @Test
    fun `getValue with fragment delegates to context`() {
        spyk(object : Fragment() {
            val string by aString
        }) {
            every { context } returns aContext
        }.string

        verify { aString(aContext) }
    }

    @Test
    fun `getValue with view delegates to context`() {
        spyk(object : View(aContext) {
            val string by aString
        }) {
            every { context } returns aContext
        }.string

        verify { aString(aContext) }
    }
}
